octavia-cli/octavia_cli/delete/commands.py

```diff
@@ -0,0 +1,91 @@
+#
+# Copyright (c) 2021 Airbyte, Inc., all rights reserved.
+#
+
+from typing import List
+
+import click
+from octavia_cli.base_commands import OctaviaCommand
+
+from .deletings import Connections, DestinationConnectorsDefinitions, Destinations, SourceConnectorsDefinitions, Sources
+
+
+@click.group("delete", help="Delete existing Airbyte resources.")
+@click.pass_context
+def _delete(ctx: click.Context):  # pragma: no cover
+    pass
+
+
+@click.group("connectors", help="Delete specific source and destination connector available on your Airbyte instance.")
+@click.pass_context
+def connectors(ctx: click.Context):  # pragma: no cover
+    pass
+
+
+@click.group("workspace", help="Delete existing resorce in workspace.")
+@click.pass_context
+def workspace(ctx: click.Context):  # pragma: no cover
+    pass
+
+
+@connectors.command(cls=OctaviaCommand, name="sources", help="Delete specific source connector available on your Airbyte instance.")
+@click.option("--sourceDefinitionId", "-i", "source_definition_id", required=True, type=str)
+@click.pass_context
+def sources_connectors(ctx: click.Context, source_definition_id: str):
+    api_client = ctx.obj["API_CLIENT"]
+    definitions = SourceConnectorsDefinitions(api_client, source_definition_id)
+    click.echo(definitions)
+    pass
+
+
+@connectors.command(
+    cls=OctaviaCommand, name="destinations", help="Delete specific destination connector available on your Airbyte instance"
+)
+@click.option("--destinationDefinitionId", "-i", "destination_definition_id", required=True, type=str)
+@click.pass_context
+def destinations_connectors(ctx: click.Context, destination_definition_id: str):
+    api_client = ctx.obj["API_CLIENT"]
+    definitions = DestinationConnectorsDefinitions(api_client, destination_definition_id)
+    click.echo(definitions)
+    pass
+
+
+@workspace.command(cls=OctaviaCommand, help="Delete existing sources in a workspace.")
+@click.option("--sourceId", "-i", "source_id", required=True, type=str)
+@click.pass_context
+def sources(ctx: click.Context, source_id: str):
+    api_client = ctx.obj["API_CLIENT"]
+    sources = Sources(api_client, source_id)
+    click.echo(sources)
+    pass
+
+
+@workspace.command(cls=OctaviaCommand, help="Delete existing destinations in a workspace.")
+@click.option("--destinationId", "-i", "destination_id", required=True, type=str)
+@click.pass_context
+def destinations(ctx: click.Context, destination_id: str):
+    api_client = ctx.obj["API_CLIENT"]
+    destinations = Destinations(api_client, destination_id)
+    click.echo(destinations)
+    pass
+
+
+@workspace.command(cls=OctaviaCommand, help="Delete existing connections in a workspace.")
+@click.option("--connectionId", "-i", "connection_id", required=True, type=str)
+@click.pass_context
+def connections(ctx: click.Context, connection_id: str):
+    api_client = ctx.obj["API_CLIENT"]
+    connections = Connections(api_client, connection_id)
+    click.echo(connections)
+    pass
```

Comment:
I have some high-level design suggestions to change these commands a bit:
```
octavia delete connector <image-name> # Or definition id if it's simpler (does not require init)
octavia delete source <source-id>  # allows users not managing resources with config to delete a source with an id (does not require init)
octavia delete destination <destination-id> # allows users not managing resources with config to delete a destination with an id (does not require init)
octavia delete connection <connection-id> # allows users not managing resources with config to delete a connection with an id (does not require init)
octavia delete --file path-to-configuration.yaml # allows user managing resources with yaml config to delete the corresponding remote resource (requires init)
```

My concern is that `octavia delete workspace` can be misleading and let user think that they will delete an Airbyte workspace... I know that you mimicked the behavior of `octavia list` (that might also deserve a bit of design change) but for this destructive command it'd love to make the UX as clearer as possible for users.

**A bit more details about `octavia delete --file path-to-configuration.yaml `**:
I think this is the most useful command as it will allow CLI users to delete a remote resource according to an existing local configuration. For safety, I would suggest that this command does not delete the YAML itself but only the remote resource. It's fine for me if you want to focus only on this command at the moment.
I think this command should accept multiple `--file` options, (you can check `octavia apply connection` for a similar implementation).

**Safety nets please 😄**: Could you please prompt the user for deletion validation before actually deleting a remote resource? The validation could be bypassed by a `--force` flag.

Reply:
First of all, thank you @alafanechere for the quick review.

I think we have to make two seperate subcommand or at least add an extra click choice option to `octavia delete connector <image-name>` for connector deletion, because it have different API's endpoint in case of [source](https://airbyte-public-api-docs.s3.us-east-2.amazonaws.com/rapidoc-api-docs.html#post-/v1/source_definitions/delete) vs [destination](https://airbyte-public-api-docs.s3.us-east-2.amazonaws.com/rapidoc-api-docs.html#post-/v1/destination_definitions/delete). Of course we can have alternative way for source/destination deletion, for example we can use single subcommand (without options) and check based on the provided definition ID is it an existing source or destination connector and call the corresponding Airbyte API endpoint. Do you prefer this alternative way or have you got some other idea to delete source/destination definition with a single subcommand?

Delete based on file (already existing) local configuration could provide an easy way for deletion. (And it is a little similar the well known device twin philosophy)
I feel that we could reorganise our code, because in the future we have got some common codes in both command (delete, apply). For example to determine the type of resource in the provided yaml configuration file. What do you think? Where shoud we put the common codes? 

I fully agree with you that safety nets is a good practise for deletion. Soon I will implement it.

Reply:
> Do you prefer this alternative way or have you got some other idea to delete source/destination definition with a single subcommand?

This is exactly the reason why we pass definition ids in `octavia generate` and not image name. But for the UX I think we should pass image's name (https://github.com/airbytehq/airbyte/issues/10981). We can call the [source_definition/list](https://airbyte-public-api-docs.s3.us-east-2.amazonaws.com/rapidoc-api-docs.html#post-/v1/source_definitions/list) and the similar destination endpoint to retrieve the definition id matching the image name. If you prefer to keep `octavia delete connector source <definition-id> ` for now it's fine for me too, but it will probably be a future improvement.

> I feel that we could reorganise our code, because in the future we have got some common codes in both command (delete, apply). For example to determine the type of resource in the provided yaml configuration file. What do you think? Where shoud we put the common codes?

I think that you can use the resources factory from `octavia-cli/octavia_cli/apply/resources.py`. With this factory, you can instantiate the resource from to the configuration path. I think that the best approach is to create a `delete` method on the object that will delete the remote twin and the local state file. You are right that if this code is reused by the delete command it does not belong to the `apply` package only. Feel free to move it to the root of the `octavia_cli` package.

> I fully agree with you that safety nets is a good practise for deletion. Soon I will implement it.
In the same PR please 🙏